lib/src/front_end/piece_factory.dart

```diff
@@ -707,6 +707,56 @@ mixin PieceFactory {
     );
   }
 
+  /// Creates a [ListPiece] for a record literal or pattern.
+  Piece createRecordCollection(
+    Token leftParenthesis,
+    List<AstNode> fields,
+    Token rightParenthesis, {
+    Token? constKeyword,
+  }) {
+    var style = switch (fields) {
+      // Record patterns with a single positional field will always
+      // have a trailing comma, like:
+      //
+      //     if (obj case (pattern,)) {
+      //       ;
+      //     }
+      [PatternField(name: null)] =>
+        const ListStyle(commas: Commas.alwaysTrailing),
+
+      // Record types with a single position field will always have a trailing
+      // comma, like:
+      //
+      //     (int,) x;
+      //
+      [var field] when field is! NamedExpression && field is! PatternField =>
+        const ListStyle(commas: Commas.alwaysTrailing),
+
+      // All other record types and patterns have regular trailing commas when
+      // split.
+      //
+      // Record types with a single named fields don't add trailing commas
+      // unless it's split, like:
+      //
+      //     ({int n}) x;
+      //
+      // Record patterns with a single named fields don't add trailing commas
+      // unless it's split, like:
+      //
+      //     if (obj case (name: value)) {
+      //       ;
+      //     }
+      _ => const ListStyle(commas: Commas.trailing)
+    };
```

Comment:
Might be a little cleaner to test for the named-single-field cases first:

```dart
var style = switch (fields) {
  // Record types or patterns with a single named field don't add a trailing
  // comma unless it's split, like:
  //
  //     ({int n}) x;
  // 
  // Or:
  //
  //     if (obj case (name: value)) {
  //       ;
  //     }
  [NamedExpression field] => const ListStyle(commas: Commas.trailing),
  [PatternField(name: _?)] => const ListStyle(commas: Commas.trailing),

  // Record types or patterns with a single positional field always have a
  // trailing comma to disambiguate from parenthesized expressions or
  // patterns, like:
  //
  //     if (obj case (pattern,)) {
  //       ;
  //     }
  [_] => const ListStyle(commas: Commas.alwaysTrailing),

  // Record types or patterns with multiple fields have regular trailing
  // commas when split.
  _ => const ListStyle(commas: Commas.trailing)
};
```

Since switches stop after the first matching case, you can often eliminate guards by reordering cases. (Whether that leads to clearer code is a matter of taste. :) )

Reply:
Ah yes, that was exactly my initial switch actually. I ended up changing it to what it is now because I didn't know if writing it out that way was weirder than writing the edge cases (single positional) and leaving everything else.

Since with writing it this way, you end up reading from the trailing cases to always trailing to trailing again.

I'm okay with changing it back to how I had it originally (which is what you're suggesting).

Reply:
Yeah, either way is fine with me.